src/include/cpu/power/scom.h

```diff
@@ -92,110 +89,123 @@ static const chiplet_id_t mcs_to_nest[] =
+ * equivalent to rscom calls for chip #0 */
+
+static inline void write_rscom(uint8_t chip, uint64_t addr, uint64_t data)
 {
 	if (addr & XSCOM_ADDR_IND_FLAG)
-		write_scom_indirect(addr, data);
+		write_scom_indirect(chip, addr, data);
 	else
-		write_scom_direct(addr, data);
+		write_scom_direct(chip, addr, data);
 }
 
-static inline uint64_t read_scom(uint64_t addr)
+static inline uint64_t read_rscom(uint8_t chip, uint64_t addr)
 {
 	if (addr & XSCOM_ADDR_IND_FLAG)
-		return read_scom_indirect(addr);
+		return read_scom_indirect(chip, addr);
 	else
-		return read_scom_direct(addr);
+		return read_scom_direct(chip, addr);
+}
+
+static inline void rscom_and_or(uint8_t chip, uint64_t addr, uint64_t and, uint64_t or)
+{
+	uint64_t data = read_rscom(chip, addr);
+	write_rscom(chip, addr, (data & and) | or);
+}
+
+static inline void rscom_and(uint8_t chip, int64_t addr, uint64_t and)
+{
+	rscom_and_or(chip, addr, and, 0);
+}
+
+static inline void rscom_or(uint8_t chip, uint64_t addr, uint64_t or)
+{
+	rscom_and_or(chip, addr, ~0, or);
+}
+
+static inline void write_rscom_for_chiplet(uint8_t chip, chiplet_id_t chiplet,
+					   uint64_t addr, uint64_t data)
+{
+	addr &= ~PPC_BITMASK(34,39);
+	addr |= ((chiplet & 0x3F) << 24);
+	write_rscom(chip, addr, data);
+}
+
+static inline uint64_t read_rscom_for_chiplet(uint8_t chip, chiplet_id_t chiplet, uint64_t addr)
+{
+	addr &= ~PPC_BITMASK(34,39);
+	addr |= ((chiplet & 0x3F) << 24);
+	return read_rscom(chip, addr);
+}
+
+static inline void rscom_and_or_for_chiplet(uint8_t chip, chiplet_id_t chiplet,
+					    uint64_t addr, uint64_t and, uint64_t or)
+{
+	uint64_t data = read_rscom_for_chiplet(chip, chiplet, addr);
+	write_rscom_for_chiplet(chip, chiplet, addr, (data & and) | or);
+}
+
+static inline void rscom_and_for_chiplet(uint8_t chip, chiplet_id_t chiplet, uint64_t addr, uint64_t and)
+{
+	rscom_and_or_for_chiplet(chip, chiplet, addr, and, 0);
+}
+
+static inline void rscom_or_for_chiplet(uint8_t chip, chiplet_id_t chiplet, uint64_t addr, uint64_t or)
+{
+	rscom_and_or_for_chiplet(chip, chiplet, addr, ~0, or);
+}
+
+/* "scom" are functions with chip number being fixed at 0 */
+
+static inline void write_scom(uint64_t addr, uint64_t data)
+{
+	return write_rscom(0, addr, data);
+}
+
+static inline uint64_t read_scom(uint64_t addr)
+{
+	return read_rscom(0, addr);
 }
 
 #if CONFIG(DEBUG_SCOM) && !defined(SKIP_SCOM_DEBUG)
 #include <console/console.h>
 
-#define write_scom(x, y)                                                \
-({                                                                      \
-	uint64_t __xw = x;                                              \
-	uint64_t __yw = y;                                              \
-	printk(BIOS_SPEW, "SCOM W %16.16llX %16.16llX\n", __xw, __yw);  \
-	write_scom((__xw), (__yw));                                     \
+#define write_scom(x, y) write_rscom(0, x, y)
+#define read_scom(x) read_rscom(0, x)
```

Comment:
Oh, I see, I thought both scom and rscom will be left, but since there is so few non-rscom accesses there is no need to keep both. In that case I'll clean it in my SCOM driver work.